### DISCUSSION TITLE
[Hele] How can I come up with such an elegant solution?

### DISCUSSION THREAD

**Discussion by RaxCruz** [cssbattle/forum#386] [General]
Hello everyone! I'm a new player

I'd like to ask how to come up with such solutions. Also, could someone help me understand this piece of code? The solution seems quite magical, but to be honest, I find it a bit challenging to comprehend. I hope to gather insights from everyone's perspectives.

![螢幕擷取畫面 2024-03-02 181747](https://github.com/cssbattle/forum/assets/39622968/8a5f971d-e283-47a2-951d-d577be13427d)

**Comment by SelenIT** [comment]
## Solution code breakdown
By default, each HTML page, even the empty one, has two visible elements, `html` and `body`. The shape of the target has a central symmetry, so a good strategy is to use each of these elements to make a half of the shape, and then invert one of the halves. The shape is made of rectangles, so multiple box-shadows of one rectangular element seem a good approach.

Here is the relevant part of the code that makes the first half of the shape (I changed the selector from `*` to `html` to make only one element cast shadows, commented our irrelevant parts, and added `outline` to highlight the actual size and the position of the element):
![Screenshot 2024-03-05 at 20 09 24](https://github.com/cssbattle/forum/assets/282754/eddfb215-9619-4bba-9f06-f647d81309f5)
The `margin: 0 45% 190` (same as `0px 45% 190px`, the lack of doctype that switches the page to [Quirks mode](https://developer.mozilla.org/en-US/docs/Web/HTML/Quirks_Mode_and_Standards_Mode) makes it possible to drop `px` for some old properties) makes the `html` element `400px - 2*0.45*400px = 400px - 360px = 40px` wide and `300px - 0 - 190px = 110px` tall, placed at the top center. The four box-shadow parts make 4 rectangles, partially overlapping one another (the first in the code makes the topmost layer, the next comes below it, and so on). The unused parts of the shadow end up offscreen.

Next, we switch the selector back to `*`. It makes the CSS right after it apply to both `html` and `body` elements, and the CSS after the second `*` — to the `body` element only (`*{*{` is a [CSS Nesting](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Using_CSS_nesting) equivalent to the `* *` selector, i.e. targets any element that has an ancestor, or any element except the root one).
![Screenshot 2024-03-05 at 20 35 01](https://github.com/cssbattle/forum/assets/282754/e6703a4b-c7c9-4367-a019-73c4a3dab645)
With `margin: 190 0 190`, we make the body element the same size as the root element, but shifted by 190px down, i.e. to the bottom center. In our case, only the top edge position of `body` matters, the bottom edge can be anywhere below 20px from the bottom window edge (in the example on your screenshot, it's offscreen 10px below it, my example shows what happens if it's 10px above it). Finally, `scale:-1` (`rotate:180deg` would work too, but is longer) inverts the bottom part, making the symmetric shape we needed.

The dot character before `<style>` (or any other virtually invisible, but not completely blank text content) is needed only to prevent the Chrome rendering bug that incorrectly handles negative margins of the empty `body` element.

The last thing is to uncomment the `background` line. In CSS, the `background` set for the `body` element gets applied to the whole viewport (unless `html` has its own `background` declaration) which is really useful in our case. In this example, the value uses the 4-hex-digit RGBA format, setting the semi-transparent color which results in the color we need after blending with default white, such colors can be found using the [Color Mixer](https://48dvyq.csb.app/) also created by @Beowolve :)

## How to come up with such solutions?
Top scorers in CSSBattle might have different strategies. For me, the 1st rule would be probably "avoid the temptation to add a new element for each part of the shape". Many shapes can be recreated with a few (or even just one) complex CSS properties, like background gradients or multiple box-shadows. For example, the shortest solution for [Target #3](https://cssbattle.dev/play/3) has only one CSS property — the `background` shorthand (although it contains all of the `background-image` — radial gradient, `background-position`, `background-size`, `background-repeat`, and `background-color` values in the compact form). So, before adding a new element to the markup, it's often useful to analyse the shape and try to recreate as much as possible of it with all the options available for existing elements, which include unlimited number of backgrounds and shadows layers, different ways to repeat the background patterns, borders (including dotted/dashed, separate side settings with properties like `border-top` and `border-block`, etc.), outlines, `border-image` (with gradients), `filter:drop-shadow`, and so on.

Hope this helps!
